### PR TITLE
Feature/78

### LINKED ISSUE #78: Parser for data ingestion
Schema: https://docs.google.com/spreadsheets/d/14u31AzaxxYfeeGY63iEEMzuSMq0DHbM2xIgz288AQ5w/edit#gid=0

- Parser for CPHATE data
- Parser for neurons
- Parser for contacts
- Parser for synapses

Yale provided the objs and xlsx files that encodes all the information required for neurons, synapses, contacts, etc.
The parser(s) will have to go through these data in order to fill the xlsx schema file attached, the xlsx will be then used by strapi to populate the db in the ingestion pipeline.

@afonsobspinto added this to ease the transition of this since Zoran will take over the development when you are in holiday:
- write documentation for the parser

### PR DESCRIPTION
**Pull request by afonsobspinto**
closes #78 

Creates 3 common parsers (spreadsheet parser, filename parser, and web parser).
Each directory/entity creates a parser that makes, most of the time, use of the common parsers (by creating a configuration file) and glues everything together with custom logic.

Current configuration files assume data  folder with [files from yale](https://drive.google.com/drive/folders/1EMiTLErFaV2CozisPtYpCv56p-U_5Ys0?usp=sharing) at the same level as the parsers folder

Mapper between neuron and wormbaseId available in the following ftp server: ftp://ftp.wormbase.org//pub/wormbase/species/c_elegans/annotation/geneIDs/c_elegans.PRJNA13758.current.geneIDs.txt.gz

Parsed files available [here](https://drive.google.com/drive/u/1/folders/1QTGEV6zXWYzGIqgbUSCN1MJqyUd2Ksuu)

Data questions available [here](https://docs.google.com/document/d/1OJYHyf9WgdIbkf0_wFBhm7eAkFHnQ8sKk2eoIsd8kFE/edit) (I think all of them are already being handled by Jerry)

### CODE REVIEW

File: ingestion/parsers/common/utils.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+class obj(object):
+  def __init__(self, d):
+    for a, b in d.items():
+      if isinstance(b, (list, tuple)):
+        setattr(self, a, [obj(x) if isinstance(x, dict) else x for x in b])
+      else:
+        setattr(self, a, obj(b) if isinstance(b, dict) else b)
+
+
+def get_args():
```
Comment (lrebscher) [inline ingestion/parsers/common/utils.py:35]: this method is not being used, want to keep it to support cli use or remove it?

File: ingestion/parsers/common/web_parser.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import requests
+from bs4 import BeautifulSoup
```
Comment (lrebscher) [inline ingestion/parsers/common/web_parser.py:2]: not listed in requirements.txt

File: ingestion/parsers/common/data_exporter.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import csv
+
+
+class DataExporter:
+  def __init__(self, data):
```
Comment (lrebscher) [inline ingestion/parsers/common/data_exporter.py:5]: your formatting doesn't conform to PEP-8, not so important for these scripts but better configure it accordingly

**Review by lrebscher — CHANGES_REQUESTED** [review]
very nice 👍 

it won't close the issue though since we need to do more adjustments once we've seen the complete dataset and folder structure

**Comment by afonsobspinto** [comment]
From @lrebscher :
1. in the parsed data I see that you merge all pre objs matching e.g. AIBR-DVC. But there exist multiple synapses between AIBR-DVC and each one will be its own instance/row. We only merge the different stages/timepoints of a neuron into one row (if that is possible)
2. Noelle confirmed that they are dyadic and triadic synapses which just means that synapse is a 1:m relationship (1-pre to m-post). This has the implication that the “post” column will become a list of neurons
3. I’m now not so sure anymore if there is a matching pair of pre and post obj for each synapse… since we don’t have the data for it yet, this can wait, but I need to get this confirmed by Noelle. Could be that there are no pairs and every obj file becomes its own synapse instance.

**Comment by afonsobspinto** [comment]
New parsed files available [here](https://drive.google.com/drive/u/1/folders/1UOaP5HUfUieZ8IgVeLqw4Srf9RbbYOVM)

[id_mapper.csv](https://github.com/MetaCell/geppetto-NeuroSCAN/files/6923606/id_mapper.csv) (promoter name -> wormbase id)

CSVs need to have well-defined headers (some were missing f.e. DCR_PromoterPipeline_Revised2021_Updated.xlsx). 
Sheet names must be consistent between files.

CPhate parser does not use the generic pipeline strategy but rather a custom script (needs to be adapted for multiple folders, delayed for when more data arrives)

Will add documentation for the parser later this week.

File: ingestion/parsers/common/tests.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+import unittest
```
Comment (lrebscher) [inline ingestion/parsers/common/tests.py:1]: thx for the test 👍

File: ingestion/parsers/promoters/parser.py [hunk 5]
```diff
@@ -0,0 +1,70 @@
+# from ingestion.parsers.common.config import Config
```
Comment (lrebscher) [inline ingestion/parsers/promoters/parser.py:1]: what about this commented code?

**Comment by lrebscher** [comment]
- add a minimal [setup.py](https://docs.python.org/3/distutils/setupscript.html#writing-the-setup-script) to ingestion folder so that one can install & run this without having to modify sys.path, IntelliJ does this automatically for you but if you run a parser from the CLI it won't be aware of the `ingestion` module. For that you better add another `ingestion` folder that will contain the source code, so `ingestion/ingestion/parser` and `ingestion/setup.py`

- it's a flexible setup but at the same time it's also quite complex compared to writing a specific parser for each entity. As you mentioned we need to add some documentation here, a general how to use and also code documentation, e.g. of the pipeline syntax and filename config

**Comment by lrebscher** [comment]
we will handle `setup.py` in a different PR

**Pull request merged**